backend/ssh.pm

```diff
@@ -0,0 +1,143 @@
+# Copyright © 2018 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package backend::ssh;
+use strict;
+use base 'backend::baseclass';
+use testapi 'get_required_var';
+use Carp 'cluck';
+
+use IO::Select;
+
+sub new {
+    my $class = shift;
+    my $self  = $class->SUPER::new;
+
+    return $self;
+}
+
+sub do_start_vm {
+    my ($self) = @_;
+
+    my $hostname = get_required_var('SSH_HOSTNAME');
+    my $password = get_required_var('SSH_PASSWORD');
+    my $username = get_required_var('SSH_USERNAME', 'root');
+
+    # truncate the serial file
+    open(my $sf, '>', $self->{serialfile});
+    close($sf);
+
+    my $ssh_console = $testapi::distri->add_console(
+        'sut',
+        'ssh_console',
+        {
+            hostname => $hostname,
+            password => $password,
+            username => $username
+        });
+    $ssh_console->backend($self);
+    $self->select_console({testapi_console => 'sut'});
+
+    # TODO use unique filename or exit if file exists
+    my $serial_file = "/dev/" . $testapi::serialdev;
+    $self->run_cmd("mkfifo $serial_file", $hostname, $password, $username);
+
+    # Listen on serial file
+    my $chan = $self->start_ssh_serial(hostname => $hostname, password => $password, username => $username);
+    $chan->exec("tail -F $serial_file");
```

Comment:
As discussed, we now have a separate code path for when we are communicating over a duplex serial connection. So there is no need to create a serial file for this purpose. You probably just want the serial file (serial0.txt) for logging kernel messages over a separate connection to the one used for sending commands to the SUT. Hopefully the cloud provider gives you a dedicate serial feed for this which doesn't require an ssh daemon to be running on the SUT, but otherwise you can figure out some workaround.

Reply:
Thx for pointing this explicitly. I will go that advice and use  also the duplex serial connection for SSH.